[tnghgks] 필터링 기능 구현

tnghgks opened the pull request:
## 구현 기능
- [x] 필터링 기능을 구현해주세요, 필터링은 특정 데이터를 하이라이트 하는 방식으로 구현해주세요
- [x] 필터링 기능은 버튼 형태로 ID값(지역이름)을 이용해주세요
- [x] 필터링 시 버튼에서 선택한 ID값과 동일한 ID값을 가진 데이터 구역만 하이라이트 처리를 해주세요
- [x] 특정 데이터 구역을 클릭 시에도 필터링 기능과 동일한 형태로 동일한 ID값을 가진 데이터 구역을 하이라이트해주세요

## 참고 사항
- Chart 관련 컴포넌트를 components 폴더안에서 따로 분리하였습니다.
- useSearchParams 훅을 사용하여 URL의 쿼리에 따라 필터의 상태가 변경되게 했습니다.
- useFilterData 커스텀 훅을 사용하여 필터 관련 데이터 조회, 추가, 초기화 기능을 분리하였습니다.
- ComposedChart 에 OnClick 이벤트를 추가하여 Bar 차트 및 Area 차트의 하이라이트 기능을 구현했습니다.
- Bar 차트의 각각 항목에 따른 Cell 컴포넌트를 렌더링하여 필터에 따라 하이라이트 기능을 구현했습니다.
- CustomDot 컴포넌트를 생성하여 Area 항목이 필터링되면 각각의 항목에 하이라이트 기능을 구현했습니다.
- 다중 필터링이 가능하게 URL의 쿼리를 배열로 받아 처리했습니다.

## SAMPLE
![차트 구현](https://user-images.githubusercontent.com/17325845/225849849-52da46a7-a5d5-4bb9-b1e9-f21dfbb5fc33.gif)

tnghgks committed:
feat: id 값 중복 제거 함수 분리

tnghgks committed:
feat: Filter Button 스타일 생성

tnghgks committed:
feat: Filter 컴포넌트 생성

tnghgks committed:
feat: CustomDot 컴포넌트 생성

tnghgks committed:
feat: 차트 클릭 또는 필터 버튼 클릭 시 필터링(하이라이트) 기능 추가

Noeyso commented on src/components/Chart/Chart.tsx:
```diff
@@ -43,19 +68,25 @@ const Chart = ({ data, start, end }: IChartProps) => {
-              fill="#868e96"
-              barSize={20}
-            />
+            <CustomDot />
```
CustomDot이 Area 컴포넌트에도 들어가있던데 여기에서 한번 더 정의한 의도가 있으신가요?

tnghgks replied:
사용법을 참고해서 따라하다보니 코드가 남은 것 같습니다! 생각 못하고 있었는데 감사합니다! 👍

hyorimcho commented on src/components/Chart/Chart.tsx:
```diff
@@ -8,15 +8,39 @@ import {
+import style from '@/components/Chart/style.module.css';
+import useFilterData from '@/lib/hooks/useFilterData';
+
+interface IActivePayload {
+  activePayload: {
+    payload: IChart;
+  }[];
+}
```
이번 과제에는 interface 분리를 안하셨네요!
컨벤션을 혼동하신걸까요🤔?

tnghgks replied:
공통적으로 사용하는 interface만 분리한다고 생각하고 있었는데 제가 컨벤션을 혼동한 것 같습니다.💦
앞으로 컨벤션에 유의하도록 하겠습니다! 감사합니다!

hyorimcho commented on src/lib/hooks/useFilterData.ts:
```diff
@@ -0,0 +1,26 @@
+import { useSearchParams } from 'react-router-dom';
+
+const useFilterData = () => {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const curFilterData = searchParams.get('filter')?.split(',') || [];
+
+  const toggleFilter = (id: string) => {
+    const isExist = curFilterData.includes(id);
+
+    if (isExist) {
+      const removedList = curFilterData.filter((item) => item !== id);
+      setSearchParams(`filter=${encodeURIComponent(removedList.join(','))}`);
+    } else {
+      curFilterData.push(id);
+      setSearchParams(`filter=${encodeURIComponent(curFilterData.join(','))}`);
+    }
+  };
+
```
지역으로 다중필터링 하는거 좋은 것 같습니다 👍🏻

greyHairChooseLife replied:
그러게요, 저도 하나씩 하는거라고만 생각했지 여러 개 할 생각은 전혀 못했어요.

Noeyso reviewed (commented):
호수님 고생하셨습니다! 플렉시스 캐릭터 ㅋㅋ 정말 상상도 못한 정체ㄴ(°0°)ㄱ

hyorimcho reviewed (commented):
코드 잘 읽었습니다. 변수명도 알아보기 쉽게 작성되어 있어 편했어요.
무엇보다도 UI에서 엄청난 호수님의 정성을 느꼈어요. 🥇 
고생하셨어요 호수님!

jhoon9494 commented on src/lib/hooks/useFilterData.ts:
```diff
@@ -0,0 +1,26 @@
+import { useSearchParams } from 'react-router-dom';
+
+const useFilterData = () => {
```
useSearchParams를 커스텀훅으로 만들어서 사용성을 높인부분 너무 좋은 것 같습니다!!

jhoon9494 reviewed (commented):
캐릭터가 인상에 남네요....ㅋㅋㅋㅋ고생하셨습니다 호수님

raw20 commented:
호수님 이번 과제도 고생많으셨습니다!! 👍👍👍 
배경에 있는 케릭터가 진짜 인상적이군요..ㅋㅋㅋ

raw20 commented on src/components/Chart/FilterBtn/FilterBtn.tsx:
```diff
@@ -0,0 +1,35 @@
+
+const FilterBtn = ({ name }: { name: string }) => {
+  const { curFilterData, toggleFilter } = useFilterData();
+  const [toggle, setToggle] = useState(false);
```
만약에 새로고침 시 선택 된 버튼들이 풀리지 않고 계속 유지가 되는건지 궁금합니다!!

jhoon9494 replied:
@raw20 

currFilterData는 url에서 값을 가져오기 때문에 새로고침하더라도 값이 유지될 것이고,
effect를 통해 curFilterData의 배열중에서 name이 포함되어있을경우 toggle을 변경시키기 때문에 유지되지 않을까 생각합니다!!

bokjunwoo commented:
호수님 캐릭터가 눈에 확들어오네요ㅋㅋㅋ😆 고생하셨습니다.

jiheon788 commented on public/assets/img/overview_bg.png:
ㅋㅋㅋㅋㅋ 캐릭터 너무 눈에들어오네요

jiheon788 reviewed (commented):
호수님! 잘봤습니다. 
다만 `from @/.../ChartFilter/ChartFilter` 현재 이런식으로 사용되고 있어서 불필요한 depth 같습니다. index.tsx로하면 `from @/.../ChartFilter` 이렇게 하는 것이 더욱 깔끔할것같습니다!

tnghgks committed:
refactor: 쿼리스트링 자원 명시를 위한 경로 분리

- 지헌님의 PR[#20] 에서 가져온 의견을 반영하였습니다.

greyHairChooseLife commented on src/lib/utils/uniqueId.ts:
```diff
@@ -0,0 +1,7 @@
+  id: string;
+}
+
+export const uniqueId = <T extends HasId>(array: T[]) => {
```
generic을 이렇게 쓰니까 사용하는 입장에서 엄청 편리하겠네요. :+1: 저는 id만 추려서 넣어주도록 했는데 ㅋㅋ;

그런데 인터페이스 추가하지않고 <T extends {id: string}>으로 생성해도 괜찮았을 것 같아요.  함수라는거 알게 해주는 이름 정도랑요..!

tnghgks committed:
refactor: useChart -> useFetch 변경 (커스텀 훅 추상화)

- 지헌님의 PR[#20] 에서 의견을 적용하였습니다.

tnghgks commented:
> 호수님! 잘봤습니다. 다만 `from @/.../ChartFilter/ChartFilter` 현재 이런식으로 사용되고 있어서 불필요한 depth 같습니다. index.tsx로하면 `from @/.../ChartFilter` 이렇게 하는 것이 더욱 깔끔할것같습니다!

tnghgks committed:
refactor: 데이터 하이라이트 시 ReferenceArea를 통한 기능 구현

- 소연님의 PR[#15] 에서 의견을 적용하였습니다.

tnghgks committed:
style: 컴포넌트 파일명을 index.tsx로 변경하여 import 경로 축소

tnghgks committed:
refactor: useFilterData => useFilterParams 로 변경 (Params key를 인자로 재사용성 증가)

greyHairChooseLife reviewed (commented):
고생하셨습니다. :+1:

tnghgks committed:
style: UI 디자인 변경

- 소연님의 PR[#15]에서 구현된 디자인을 적용했습니다.
- 효림님의 PR[#17]에서 구현된 색상을 적용했습니다.
- 다만 하이라이트 시 기존 색상이 구분 가지 않아 수정했습니다.

tnghgks committed:
fix: 차트 이름 수정

jiheon788 commented on src/lib/hooks/useFetch.ts:
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useState } from 'react';
+import apiClient from '@/api';
+
+const useFetch = <T>(
+  defaultValue: T,
+  url: string,
+  preProcessData: (data: T) => void,
+): [T, boolean, boolean] => {
+  const [payload, setPayload] = useState(defaultValue);
+  const [isLoading, setIsLoading] = useState(false);
+  const [isError, setIsError] = useState(false);
+
+  useEffect(() => {
+    setIsLoading(true);
+    apiClient
+      .get(url)
+      .then((res) =>
+        preProcessData === null
+          ? res.data.response
+          : preProcessData(res.data.response),
+      )
+      .then((res) => setPayload(res))
+      .catch((_) => setIsError(true))
+      .finally(() => setIsLoading(false));
+  }, [preProcessData, url]);
+
+  return [payload, isLoading, isError];
+};
+
+export default useFetch;
```
이거 문제 안생기나요?  `preProcessData`는 원본데이터를 전처리 해주는 거라 반환값이 `T`입니다.  즉 `(data: any) => T` 이렇게 해야 맞을거같습니다.
추가로 전처리 없는 경우 가정해서 기본값을 null로 해놓고 18번째 라인에 null일경우 그대로 반환, null이 아닐경우 전처리 함수 후 setState를 합니다. 현재는 전처리함수가 무조건있으니 18 19번라인의 삼항연산자가 없어도 될듯합니다!

tnghgks replied:
말씀하신대로 전처리 함수가 없을 경우를 생각해서 null 을 넣는게 맞는것 같습니다!
` preProcessData: ((data: T) => T) | null = null ` 이와같이 유니언으로 타입을 처리하고 디폴트값으로 null 을 다시 설정하였습니다. 그렇기 때문에 18,19번라인의 삼항연산자는 변경 없이 그대로 두었습니다!

jiheon788 commented:
@tnghgks  

추가로 바뀐 샘플 확인하였습니다. 하이라이트의 목적은 보고 싶은 데이터를 더 잘보기 위함인데 레퍼런스차트의 색감이 너무쎄서 기존 데이터가 묻히는거 같습니다. 비유하자면 형광펜을 써서 밑줄을 그엇더니 글자가 안보이는 느낌입니다! 조금 약한 색감을 사용해주시는게 좋아보입니다.

![image](https://user-images.githubusercontent.com/90181028/225826212-85ee68c1-7e21-4c75-834e-4b05b25f05fc.png)

tnghgks committed:
fix: preProcessData 함수 타입 변경 및 디폴트 값(null) 추가

tnghgks committed:
style: highlight 색상 변경

hyorimcho commented:
고생하셨습니다 호수님 👍🏻
